File: 451-sort-characters-by-frequency/sort-characters-by-frequency.cpp
#include <bits/stdc++.h>
using namespace std;


class Solution {
public:
    string frequencySort(string s) {
          unordered_map <char,int> mpp;
          for(int i=0;i<s.size();i++)
          {
             mpp[s[i]]++;
          }
          vector<vector<int>> arr;   
          for(auto it:mpp)
          { 
            cout<<it.first<<it.second;
            arr.push_back({it.second,it.first});
          }


          sort(arr.begin(),arr.end(), greater<vector<int>>());

          for(int i=0;i<arr.size();i++)
          {
            for(int j=0;j<arr[i].size();j++)
            {
                cout<<arr[i][j];
            }
            cout<<endl;
          }

           string ans;
          for(int i=0;i<arr.size();i++)
          {
            while(arr[i][0]!=0)
              { 
                ans.push_back(arr[i][1]);
                arr[i][0]--;
              }
          }

          return ans;
    }
};

File: 451-sort-characters-by-frequency/sort-characters-by-frequency.kt
class Solution {
    fun frequencySort(s: String): String {
        val arr = IntArray(size = 128) { 0 }
        val pQueue = PriorityQueue<Pair<Int, Int>>(compareByDescending { it.second })
        s.forEach { arr[it.toInt()]++ }
        
        arr.forEachIndexed { index, i -> 
            pQueue.offer(Pair(index, i))
        }
        
        val strOut = StringBuilder()
        while (pQueue.isNotEmpty()) {
            val (charInt, freq) = pQueue.poll()
            repeat(freq) {
                strOut.append(charInt.toChar())
            }
        }

        return strOut.toString()
    }
}